### PR TITLE
msedgedriver: init at 130.0.2849.1

### PR DESCRIPTION
**Pull request by Daholli**
This initializes a package from 

[https://developer.microsoft.com/en-us/microsoft-edge/tools/webdriver?form=MA13LH](https://developer.microsoft.com/en-us/microsoft-edge/tools/webdriver?form=MA13LH)

for now this is only the stable channel, and would need to be updated together with the `microsoft-edge` package I would love to include the edgedrivers for the other channels aswell but I don't know how yet

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Daholli** [comment]
would close #192050

**Comment by Daholli** [comment]
this exists in multiple channels and for multiple platforms but I need help / some guidance as to how to start the others

**Comment by Daholli** [comment]
@zanculmarktum @kuwii @rhysmdnz 

can you guys maybe help me setting up the other channels? Since this one only works with the `microsoft-edge` package, trying to copy the way it was set up for the browser was not working at all.

**Comment by Daholli** [comment]
waiting for #314934 so I am not getting merge conflicts for the maintainers list

File: pkgs/by-name/ms/msedgedriver/package.nix
Comment (gepbird) [inline pkgs/by-name/ms/msedgedriver/package.nix:24]: `channel` is not used anywhere, can be deleted.
`hash` is only used at one place, it can be moved there.
`version` is used multiple times. For builders like `stdenv.mkDerivation`, the `finalAttrs` pattern is preferred over `let-in`.
```suggestion

stdenv.mkDerivation (finalAttrs: {
  pname = "msedgedriver";
  version = "129.0.2792.91";

  src = fetchurl {
    url = "https://msedgedriver.azureedge.net/${finalAttrs.version}/edgedriver_linux64.zip";
    hash = "sha256-69+quFNgjioVBtsRSZDPBYXWhtCvX6cZ5TqGhhf5Ht8=";
  };
```

File: pkgs/by-name/ms/msedgedriver/package.nix
Comment (gepbird) [inline pkgs/by-name/ms/msedgedriver/package.nix:55]: Closing parenthesis for `finalAttrs`
```suggestion
})
```

File: pkgs/by-name/ms/msedgedriver/package.nix
Comment (gepbird) [inline pkgs/by-name/ms/msedgedriver/package.nix:48]: Please set these fields.

File: pkgs/by-name/ms/msedgedriver/package.nix
Comment (gepbird) [inline pkgs/by-name/ms/msedgedriver/package.nix:52]: There's no aarch64 binary in the zip, only x86_64.
```suggestion
    platforms = [ "x86_64-linux" ];
```

**Review by gepbird — CHANGES_REQUESTED** [review]
Thanks for your contribution!

Since evaluation fails due to your maintainer init commit only being in your other PR, please mark it as a draft to avoid accidental merges (it has happened before). You can either wait for that to get merged or include the same commit in this PR too.

Please fix the PR title and commit message, it should be `msedgedriver: init at 129.0.2792.91`.

Also don't advertise your PR there while it hasn't been reviewed, use https://discourse.nixos.org/t/prs-ready-for-review next time.

**Comment by Daholli** [comment]
> Also don't advertise your PR there while it hasn't been reviewed, use https://discourse.nixos.org/t/prs-ready-for-review next time.

yeah sorry i read it afterwards and deleted it, but then saw that the deletion didn't actually delete the message but rather removed the text

**Comment by gepbird** [comment]
> this exists in multiple channels and for multiple platforms but I need help / some guidance as to how to start the others

To make it work on different architectures, you can use something similar to how it is done with other binary packages like [dbeaver-bin](https://github.com/nixos/nixpkgs/blob/master/pkgs/by-name/db/dbeaver-bin/package.nix#L22-L42).

I'm not sure what's the best way for channels in `by-name`, a possibility is including a `channel` parameter in the input set (where `lib`, `stdenvNoCC`... live) and making `callPackage`s for each channel in `top-level/all-packages.nix`:
```nix
  msedge = callPackage ../by-name/ms/msedge/package.nix { channel = "stable"; };
  msedge-beta = callPackage ../by-name/ms/msedge/package.nix { channel = "beta"; };
```
I'll check if this is accepted for `by-name` later.

**Comment by rhysmdnz** [comment]
If you find an answer to how to do channels nicely I could be interested.

For msedge I'm currently looking at following google-chrome and not packaging the beta channels. (#347836)

Does the version of this need to exactly match edge? Should we tie this to the edge update script?

**Comment by Daholli** [comment]
> If you find an answer to how to do channels nicely I could be interested.
> 
> For msedge I'm currently looking at following google-chrome and not packaging the beta channels. (#347836)
> 
> Does the version of this need to exactly match edge? Should we tie this to the edge update script?

I think the driver needs to be >= edge to work and only works within the same channel, have to read up on that tho

**Comment by Daholli** [comment]
@rhysmdnz According to [microsoft](https://learn.microsoft.com/en-us/microsoft-edge/webdriver-chromium/?tabs=c-sharp) only the first three parts of the four-part version number must match, so if that changes in edge we also need to update this

**Comment by Daholli** [comment]
```
[61173:61228:1016/171349.318046:ERROR:edge_encryption_keys_manager_aad.cc(542)] Failed to initialize EncryptionKeysProtector.
[61173:61228:1016/171349.318061:ERROR:edge_encryption_keys_manager_aad.cc(186)] Got invalid encryption keys!
[61173:61173:1016/171357.684112:ERROR:CONSOLE(1)] "Uncaught (in promise) TypeError: Failed to fetch", source: devtools://devtools/bundled/models/live-metrics/live-metrics.js (1)
[61173:61230:1016/171358.549031:ERROR:http_transport_libcurl.cc(118)] dlopen:libcurl.so: cannot open shared object file: No such file or directory
[61173:61230:1016/171358.549043:ERROR:http_transport_libcurl.cc(118)] dlopen:libcurl-gnutls.so
.4: cannot open shared object file: No such file or directory
[61173:61230:1016/171358.549046:ERROR:http_transport_libcurl.cc(118)] dlopen:libcurl-nss.so.4: cannot open shared object file: No such file or directory
[61173:61230:1016/171358.549049:ERROR:http_transport_libcurl.cc(118)] dlopen:libcurl.so.4: cannot open shared object file: No such file or directory
[1016/171358.561423:ERROR:elf_dynamic_array_reader.h(64)] tag not found
[1016/171358.561506:ERROR:process_memory_range.cc(75)] read out of range
[1016/171358.567211:ERROR:http_transport_libcurl.cc(118)] dlopen:libcurl.so: cannot open shared object file: No such file or directory
[1016/171358.567217:ERROR:http_transport_libcurl.cc(118)] dlopen:libcurl-gnutls.so.4: cannot open shared object file: No such file or directory
[1016/171358.567219:ERROR:http_transport_libcurl.cc(118)] dlopen:libcurl-nss.so.4: cannot open shared object file: No such file or directory
[1016/171358.567221:ERROR:http_transport_libcurl.cc(118)] dlopen:libcurl.so.4: cannot open shared object file: No such file or directory
[1016/171358.568002:ERROR:process_memory_range.cc(75)] read out of range
[1016/171358.569808:ERROR:http_transport_libcurl.cc(118)] dlopen:libcurl.so: cannot open shared object file: No such file or directory
[1016/171358.569817:ERROR:http_transport_libcurl.cc(118)] dlopen:libcurl-gnutls.so.4: cannot open shared object file: No such file or directory
[1016/171358.569819:ERROR:http_transport_libcurl.cc(118)] dlopen:libcurl-nss.so.4: cannot open shared object file: No such file or directory
[1016/171358.569821:ERROR:http_transport_libcurl.cc(118)] dlopen:libcurl.so.4: cannot open shared object file: No such file or directory
```

was playing around with this a little bit today and when opening the dev tools on edge my browser crashed @rhysmdnz 
is that something that happens on your end too?

**Comment by rhysmdnz** [comment]
Ooohh, yes. Seems to be fixed by #347836 adding curl.

**Comment by Daholli** [comment]
fixed merge conflicts

File: pkgs/by-name/ms/msedgedriver/package.nix
Comment (gepbird) [inline pkgs/by-name/ms/msedgedriver/package.nix:20]:
```suggestion
  src = fetchzip {
    url = "https://msedgedriver.azureedge.net/${finalAttrs.version}/edgedriver_linux64.zip";
    hash = "sha256-t/faImp+nUpyELQPN+X/g54bTwG2aZU6S+4BClWe3aY=";
    stripRoot = false;
  };
```

File: pkgs/by-name/ms/msedgedriver/package.nix
Comment (gepbird) [inline pkgs/by-name/ms/msedgedriver/package.nix:3]:
```suggestion
  fetchzip,
```

File: pkgs/by-name/ms/msedgedriver/package.nix
Comment (gepbird) [inline pkgs/by-name/ms/msedgedriver/package.nix:37]: No longer necessary, the zip file is automatically extracted
```suggestion
```

File: pkgs/by-name/ms/msedgedriver/package.nix
Comment (gepbird) [inline pkgs/by-name/ms/msedgedriver/package.nix:32]:
```suggestion
```

File: pkgs/by-name/ms/msedgedriver/package.nix
Comment (gepbird) [inline pkgs/by-name/ms/msedgedriver/package.nix:9]:
```suggestion
```

File: pkgs/by-name/ms/msedgedriver/package.nix
Comment (gepbird) [inline pkgs/by-name/ms/msedgedriver/package.nix:23]: It looks unnecessary, any specific reason why it was added?
```suggestion
```

File: pkgs/by-name/ms/msedgedriver/package.nix
Comment (gepbird) [inline pkgs/by-name/ms/msedgedriver/package.nix:25]:
```suggestion
    libxcb
```

File: pkgs/by-name/ms/msedgedriver/package.nix
Comment (gepbird) [inline pkgs/by-name/ms/msedgedriver/package.nix:10]:
```suggestion
  libxcb,
```

**Review by gepbird — CHANGES_REQUESTED** [review]
As previously mentioned, after https://github.com/NixOS/nixpkgs/pull/348821 we only need to provide a stable msedgedriver (and therefore we can also keep the derivation in by-name).

What about making it work with different architectures? Again feel free to ask for help when you get stuck, and a reference for that is in [dbeaver-bin](https://github.com/nixos/nixpkgs/blob/master/pkgs/by-name/db/dbeaver-bin/package.nix#L22-L42).

File: pkgs/by-name/ms/msedgedriver/package.nix
Comment (Daholli) [reply pkgs/by-name/ms/msedgedriver/package.nix:23]: I copied the dependecies of the chromium driver since the two are basically equivalent

Comment (Daholli) [reply pkgs/by-name/ms/msedgedriver/package.nix:23]: lets see if it works without it

Comment (Daholli) [reply pkgs/by-name/ms/msedgedriver/package.nix:23]: it is still starting without it and seems to have all dependencies, ill remove it

**Comment by Daholli** [comment]
> As previously mentioned, after #348821 we only need to provide a stable msedgedriver (and therefore we can also keep the derivation in by-name).
> 
> What about making it work with different architectures? Again feel free to ask for help when you get stuck, and a reference for that is in [dbeaver-bin](https://github.com/nixos/nixpkgs/blob/master/pkgs/by-name/db/dbeaver-bin/package.nix#L22-L42).

I think I have a first attempt, stupid question, but how do i get the hashes / test if it works if i do not have a mac on hand?

**Comment by Daholli** [comment]
i wish this was part of the documentation:
```nix
nix-prefetch '{ stdenv, fetchzip }: stdenv.mkDerivation rec { name = "test"; src = fetchzip { url = https://msedgedriver.azureedge.net/130.0.2849.1/edgedriver_mac64_m1.zip; stripRoot = false; }; }'
```

**Comment by Daholli** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 348821`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>msedgedriver</li>
  </ul>
</details>

File: pkgs/by-name/ms/msedgedriver/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/ms/msedgedriver/package.nix:52]: Please invert this condition to make things simpler

File: pkgs/by-name/ms/msedgedriver/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/ms/msedgedriver/package.nix:72]:
```suggestion
    description = "WebDriver implementation that controls an Edge browser running on the local machine";
```

**Review by gepbird — APPROVED** [review]
Tested on x86_64-linux, LGTM!

**Comment by gepbird** [comment]
Since https://github.com/NixOS/nixpkgs/pull/314934 got merged, you can drop the maintainer commit from this PR. It won't cause issues if you don't do it, just the git log will look nicer.

**Review by gepbird — APPROVED** [review]

**Review by JohnRTitor — COMMENTED** [review]
We really should have an update script for this.

Perhaps you can take inspiration from https://github.com/NixOS/nixpkgs/blob/98b70abd0b12662a2ff64ddc2beec9b6f924d6a4/pkgs/by-name/wa/waveterm/update.sh and https://github.com/aucub/nixpkgs/blob/98b70abd0b12662a2ff64ddc2beec9b6f924d6a4/pkgs/by-name/wa/waveterm/package.nix#L38-L60 which implements a similar multi arch update system.

File: pkgs/by-name/ms/msedgedriver/package.nix
Comment (gepbird) [inline pkgs/by-name/ms/msedgedriver/package.nix:16]: An update script would be nice, but this issue makes it harder to do that.

Comment (JohnRTitor) [reply pkgs/by-name/ms/msedgedriver/package.nix:16]: Do we have an API to query versions of MS EDGE perhaps?

**Pull request merged**
